cmd/keymasterd/config.go

```diff
@@ -377,55 +394,79 @@ func (state *RuntimeState) loadSignersFromPemData(signerPem, ed25519Pem []byte)
 	return nil
 }
 
-func (state *RuntimeState) loadExternalSigners() error {
-	state.logger.Debugf(3, "Top of loadExternalSigners")
-	var signer crypto.Signer
-	switch state.Config.Base.ExternalSignerConf.Type {
+func (sconfig *ExternalSignerConfig) Parse() (*ParsedExternaSignerConfig, error) {
+	var parsedConfig ParsedExternaSignerConfig
+	switch sconfig.Type {
 	case "yubipiv":
-		state.logger.Debugf(3, "loadExternalSigners yubipiv branch")
-		//parse as url
-		parsedYK, err := url.Parse(state.Config.Base.ExternalSignerConf.Location)
+		parsedConfig.Type = ExternalSignerYubiPIV
+		parsedYK, err := url.Parse(sconfig.Location)
 		if err != nil {
-			return err
+			return nil, fmt.Errorf("Cannot parse url for yubipiv")
 		}
 		serial, err := strconv.ParseUint(parsedYK.Host, 10, 32)
 		if err != nil {
-			return err
+			return nil, fmt.Errorf("Invalid Host name '%s' is not converable to serial", parsedYK.Host)
 		}
-		pin := "123456" //this is the yubikey default pin
-		var pubKey any
-		pubKey = nil
-		if parsedYK.User != nil {
-			pass, ok := parsedYK.User.Password()
-			if ok {
-				pin = pass
-			}
-			b64derPubKey := parsedYK.User.Username()
-			state.logger.Debugf(3, "loadExternalSigners der pub=%s", b64derPubKey)
-			derPubKey, err := base64.URLEncoding.DecodeString(b64derPubKey)
-			if err != nil {
-				return fmt.Errorf("Invalid pub key encoding err=%s", err)
-			}
-			pubKey, err = x509.ParsePKIXPublicKey(derPubKey)
-			if err != nil {
-				return fmt.Errorf("Invalid pub key err=%s", err)
-			}
-		} else {
-			state.logger.Debugf(0, "Notice: Yubikey using default pin")
+		parsedConfig.YKSerial = uint32(serial)
+		parsedConfig.PIVPin = "123456"
+		if parsedYK.User == nil {
+			parsedConfig.PublicKey = nil
+			return &parsedConfig, nil
+		}
+		b64derPubKey := parsedYK.User.Username()
+		derPubKey, err := base64.URLEncoding.DecodeString(b64derPubKey)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid pub key encoding err=%s", err)
 		}
+		parsedConfig.PublicKey, err = x509.ParsePKIXPublicKey(derPubKey)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid pub key err=%s", err)
+		}
+		pass, ok := parsedYK.User.Password()
+		if ok {
+			parsedConfig.PIVPin = pass
+		}
+		return &parsedConfig, nil
+	case "AWS-kms":
```

Comment:
I think it would be cleaner to make this just `AWS`. We may want to support CloudHSM.